proposals/0000-custom-type-errors.rst

```diff
@@ -0,0 +1,188 @@
+.. proposal-number:: 
+
+.. trac-ticket:: 
+
+.. implemented:: 
+
+.. highlight:: haskell
+
+This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/59>`_.
+
+.. contents::
+
+Custom type errors
+==================
+
+This proposes the addition of two combinators over ``Constraint``, namely ::
+
+    IfUndischarged :: Constraint -> ErrorMessage -> Constraint
+    IfApart        :: Type -> Type -> Constraint -> Constraint -> Constraint
+
+which enable library authors to provide custom error messages for their libraries.
+
+In addition, this proposal also motivates changing the constraint solving order to prioritize arguments to functions over the check of the application itself.
```

Comment:
Since solving is non-deterministic, we are not sure about which `IfApart` constraints are solved before others. This would introduce a bit of determinacy so that we could know, for example, that when we check some argument for apartness, all constraint solving for this argument has already been done. That way, we have a higher chance of detecting an apartness condition instead of just falling back to the "don't know" case.